Ported update.sh to D which will make it easier to run on Windows/OS X also

Laeeth opened the pull request:
Hi.

Ported update.sh to D - hope this might be of some use.  I don't pretend it is pretty generally, and in particular I tried to stick to the structure of the original script for clarity, even though it would be more elegant to factor out the repetition.

Just run makeit.sh to compile - you may want to change the compiler to dmd.  I destroyed my system dmd with a null file when working on it, so changed to ldc temporarily.


Laeeth.

andralex commented:
This requires some additional work:

* add building the binaries to the build (*.mak)
* create pages on the website for explaining what it does and downloading it (this is a logistics matter you could take with @CyberShadow)

There's also significant tactical work you need to do on the sources. I'll make some notes.

andralex commented on updatesh_d/scriptutil.d:
```diff
@@ -0,0 +1,184 @@
+module scriptutil;
+import core.sys.posix.unistd:geteuid;
```
spaces around `:`

andralex commented on updatesh_d/scriptutil.d:
```diff
@@ -0,0 +1,184 @@
+module scriptutil;
+import core.sys.posix.unistd:geteuid;
+static import std.ascii;
+import std.typecons:Tuple,tuple;
```
space after comma

andralex commented on updatesh_d/scriptutil.d:
```diff
@@ -0,0 +1,184 @@
+import std.algorithm:max;
+import std.random:uniform;
+import std.conv:to;
+import std.stdio:writefln,stdout;
```
try to make most imports local

andralex commented on updatesh_d/makeit.sh:
```diff
@@ -0,0 +1 @@
+ldc2 dlangorg/updatesh.d dlangorg/scriptutil.d
```
ldc2 is less likely to be present than dmd. Anyhow this must go in the makefiles.

andralex commented on updatesh_d/scriptutil.d:
```diff
@@ -0,0 +1,184 @@
+// shamelessly nicked from http://www.mktemp.org/repos/mktemp/file/e4089ef40fd0/mkdtemp.c
+
+
+private string makeHash(immutable string path)
```
`immutable` is a bit redundant here and would be appropriate mostly for large functions, just use `in` or nothing

andralex commented on updatesh_d/scriptutil.d:
```diff
@@ -0,0 +1,184 @@
+	return digest!CRC32(path).crcHexString ~ digest!CRC32((Clock.currTime.toISOExtString)).crcHexString;
+}
+
+private enum permittedCharacters= std.ascii.letters ~ std.ascii.digits;
```
spaces around `=`

andralex commented on updatesh_d/scriptutil.d:
```diff
@@ -0,0 +1,184 @@
+private string makeRandomAlphaNumeric(size_t numChars)
+{
+	string ret="";
+	foreach(i;0..numChars)
```
space after `;`

andralex replied:
spaces around `..`

andralex commented on updatesh_d/scriptutil.d:
```diff
@@ -0,0 +1,184 @@
+	foreach(i;0..numChars)
+	{
+		auto r=uniform(0,permittedCharacters.length);
+		ret~=to!char(r);
```
spaces around `~=` etc

andralex commented on updatesh_d/scriptutil.d:
```diff
@@ -0,0 +1,184 @@
+	createDirectory,
+}
+
+enum MaxTries=100_000;
```
Group constants at the top (as opposed to mixing them with function definitions). Their order doesn't matter anyway and it's good to have all configuration constants in a place.

andralex replied:
value constants start with lowercase

andralex commented on updatesh_d/scriptutil.d:
```diff
@@ -0,0 +1,184 @@
+	}
+	
+	throw new Exception ("makeTempEntry unable to  "~ to!string(mode) ~"  in generic path " ~ path );
+	assert(0);
```
remove

andralex commented on updatesh_d/updatesh.d:
```diff
@@ -0,0 +1,423 @@
+{
+	 writefln(
+ 	"updatesh.d - for the D programming language
+	\n
```
So is it \n or just the carriage returns embedded in the string? Choose one.

andralex commented:
that's what I got for now

Laeeth commented:
Thanks, Andrei.

I am honoured to be destroyed, and will make the changes you suggest.

Laeeth.

Sent from my iPad

> On 29 Jan 2015, at 18:59, Andrei Alexandrescu <notifications@github.com> wrote:
> 
> that's what I got for now
> 
> —
> Reply to this email directly or view it on GitHub.
>

Laeeth commented:
Btw just as some background - I started learning to code when I was ten (1983)  but pursued hedge fund money management as a career.  I need to refresh my skills to develop some big data ideas.  So many of the developments that took place whilst I was busy doing other things passed me by, and I have spent the past year as a sabbatical catching up a little.  But practice makes better, and so does good critique and I thank you for the latter.


Laeeth.

Sent from my iPad

> On 29 Jan 2015, at 18:53, Andrei Alexandrescu <notifications@github.com> wrote:
> 
> This requires some additional work:
> 
> add building the binaries to the build (*.mak)
> create pages on the website for explaining what it does and downloading it (this is a logistics matter you could take with @CyberShadow)
> There's also significant tactical work you need to do on the sources. I'll make some notes.
> 
> —
> Reply to this email directly or view it on GitHub.
>

andralex commented:
Awesome, thanks. FWIW your style is totally fine, just a bit different from the usual in this codebase.

Laeeth commented:
Thanks.  I incorporated your suggestions - hope that is okay now.  I haven't yet been able to test on Windows as I am rebuilding my VM and might be a couple of days before I have time.  So in the meantime I changed posix.mak but didn't want to touch the windows makefile in case doing so broke the whole build.

CyberShadow commented:
Before too much work is expended here, I'm going to plug my project again:

https://github.com/CyberShadow/Digger

It does almost all that this small script does, and much more. Crucially, it obtains *all* the dependencies needed to build D on Windows - even Visual Studio components for Win64 builds - and installs them in a directory, without affecting the host system.

There is also DVM, which doesn't download D but knows how to build and install downloaded D source.

I know Andrei knows about Digger, and I've posted a link to the forum thread, so I'd like to question the reason for spending efforts on solved problems.

Laeeth commented:
I had a chat with Vladimir on irc earlier.  I didn't know about Digger before, and it does look nice, and it is easy enough to use.

I don't think it's a terrible thing to have choices in build/fetch tools, and it is good to have the option of a minimal tool because less brittle and easy to fix.  (Nassim Taleb's antifragility).

But I will leave it to you.  A basic README.MD is in the repo, although my markup knowledge is very basic, and it needs a slight tidyup.  The other changes you asked for have been incorporated.

marler8997 commented:
@CyberShadow I'm putting the nightly build together for asm.dlang.org.  I used update.sh to download and update and build but I can switch to Digger.  I have tried to use digger before but I really didn't like using a browser to connect to it.  I believe you can also use the command line but that kinda turned me off so I moved on to other things.  However, I do want the setup the nightly build the right way so I'm asking for your advice on this.  Which tool should we use?  update.sh? Digger? Dvm? and why?  Thanks.

marler8997 commented on FetchDlang/fetchdlang.d:
```diff
@@ -0,0 +1,475 @@
+
+// Take care of the command line arguments
+    writefln("* fetchdlang: a tool for building and maintaining the dlang toolchain from github");
+    foreach(arg; args[1 .. $])
```
Were you not able to use std.getopt for this?

Laeeth replied:
honestly, as I wrote it, I sort of clenched my teeth and forced myself to choose the path of a line of line translation rather than a  write the functionality from scratch path.  if I were writing from scratch, I would also have made as much as possible data rather than hard coding it - so you have a json table of what you need to do.  but it is harder to understand, and harder to see what it is doing that way.

On the one hand, the reason to have done this is because it is appealing to have a native code tool and to eat your own dogfood.  On the other everything that goes in here is somewhat a showcase, so it makes sense to do things as elegantly as possible.

I am happy to change if people feel strongly about it.

Wrt digger vs a simple little tool, I wasn't aware of digger, and it is beautiful, like Vladimir's other projects.  It is a philosophical choice about simple tools that do one thing versus powerful tools that unavoidably are more complex and intimidating.  I do not think it hurts to give people a choice, because of the compounding of complexity problem when you are new to a domain.  I think thtat's why Andrei was excited about update.sh - his focus being to make it easy and unintimidating for new developers to get involved, and because small initial frictions have large consequences.

Anyway - I leave it to you to decide.


Laeeth.

CyberShadow commented:
> have tried to use digger before but I really didn't like using a browser to connect to it. I believe you can also use the command line but that kinda turned me off

The web interface is a completely separate program built on top of the command-line "core".

> However, I do want the setup the nightly build the right way so I'm asking for your advice on this. Which tool should we use? update.sh? Digger? Dvm? and why? Thanks.

At last year's DConf, Andrei told me we should get Digger into the D-P-L.org repository. If he still means it, then it could probably replace update.sh.

marler8997 commented:
Good to know the web interface is separate.  I'll look into it again.

CyberShadow commented:
@marler8997 BTW, would you be interested in looking at DPaste as well? It is in dire need of a maintainer, and I have my hands full already unfortunately.

marler8997 commented:
Sure.  Email me the details, what needs to be done.

CyberShadow commented:
Sure:

> This may seem surprising but rationale is very simple - there is no scripting language that would be guaranteed to be installed on any operating system. 

Also people will need a D compiler to build D anyway (ddmd).

> It doesn't try to always track the master HEAD of submodule repositories - sadly, git does not allow to use branches-submodules without saving specific commit hash. 

Git 1.8.2 added the ability to track branches in submodules.

> You may notice that after the build all binaries get placed into ./bin folder inside the cloned repo with a -dev suffix. 

This may make using tools such as rdmd or dub difficult. Something like what dvm does might be better. Right now I'm working on an "install" command for digger, alongside an "uninstall" command - with caching, you could use it as a dvm replacement. But for a simple "just get me on my feet" tool, this might be the simplest solution.

> Integration with Digger is planned for advanced traversal through D-Programming-Language repository history (bisection, for example)

Care to go into more detail here?

CyberShadow commented:
> Only difference with "normal" submodule seems that it knows remote tracking branch and thus git submodule update --remote can be used to keep moving to current master.

I see. I haven't used it myself.

> I don't see how. It provides full copy of the toolchain with -dev suffixes, including rdmd-dev and dub-dev - this is layout I have been using personally for few years now to do DMD/Phobos development.

So rdmd-dev and dub-dev are patched to run dmd-dev instead of dmd?

> At the very least - have digger as one of submodules / move it to D-Programming-Language/tools and provide wrapper commands to top-level scripts that call it using the same sandbox layout. I was actually hoping you could spot more opportunities for Digger from such standardized layout :)

Well, for one, I think it wouldn't be hard to implement this script's functionality *on top of* Digger, which shouldn't be very difficult, and which would give it Digger's advantages (caching, getting all dependencies for Win32/Win64, etc.)

andralex commented:
With apologies to @Laeeth, I think we should double down on Digger instead of diverting into creating and maintaining a new tool. `tools/update.sh` does remain a fine option for people who can't use Digger because they don't have any installation of D and just want to get started.

I knew about Digger but never used it so I neglected its relevance to this. Apologies.

I filed https://issues.dlang.org/show_bug.cgi?id=14091. @Laeeth @CyberShadow if there is agreement on the course of action please close this. Thanks.

Laeeth commented:
It certainly makes sense to build on what we have, and it is normal to have experiments that one decides against using.  So I shall close this pull request.

CyberShadow commented:
> because they don't have any installation of D and just want to get started.

Incidentally, I've started creating binary packages of Digger for all major platforms:
https://github.com/CyberShadow/Digger/releases/tag/1.1-alpha-1

CyberShadow commented:
@ibuclaw was dealing with the same question while creating Debian/Ubuntu packages just a few days ago. I believe he patched the code to move the "www" directory somewhere else.

Let me know if I can help.
